initial commit for EntityInfraWalletCard

nia-potato opened the pull request:
Hi @gluckzhang @emillg 

This PR Adds `EntityInfraWalletCard` 
Closes https://github.com/electrolux-oss/infrawallet/issues/44

this adds a sample entityCard to respective components for gcp-projects-ids in particular. 

it fetches the gcp-project-id in two ways currently: 
- if a entity contains the annotation of `infrawallet.io/gcp-project-ids: gcp-project-idA, gcp-project-idB`
- if a components contains a depnendsOn entity that is `Kind: Resource` and `Type: gcp-project`  ( this method is a more native method to handle gcp-projects in backstage, it assumes the user has already import their organizations gcp projects into backstage catalog, and the user can directly link their systems and components to use dependsOn : gcp-project so that a seperate annotation is not needed. This is good for there are already scaffolder workflows that auto map gcp project when provisioning new services, cost related data will automatically be added on without live patching a catalog-info.yaml file and finding its related gcp-project. 



As you can see, I currently have a very basic mockup just for debugging purposes, thus there isnt too much going on the UI. 

the query i sent to the backend is 

```

"GET /api/infrawallet/reports?&filters=(gcp_project_id=backstage-dev)&tags=()&groups=&granularity=monthly&startTime=1724343492212&endTime=1727021892212 HTTP/1.1" 200 - "http://localhost:3000/"
```

but as you can see from the UI, it resolved a cost info per service (for pretty much every gcp-project id) 

im assuming the backend isnt setup to just handle query of just a specific gcp cost data and is made available for the api client to use yet. How do you suggest i move forward? Should i be using another existing API instead or just build one specific for gcp? 


also the current flow seems to trigger a new job per each entity that has a gcp project. I think that is uneconomical , can i re-use the first job being ran via the main infrawallet plugin api call, re-use that data for every entity, then only selectively update per entity data only when the main plugin get an update? If you agree with such method, please point me to right place you would recommend me to add it. 

<img width="732" alt="Screenshot 2024-09-23 at 00 38 35" src="https://github.com/user-attachments/assets/fff8e983-ce7f-47cf-bc69-010fc81e920a">

gluckzhang commented:
Hi @nia-potato, thank you for working on integrating InfraWallet with Catalog! 🚀  I see how you are going to show the costs of GCP projects now, for `EntityInfraWalletCard`, I think it can be more generic. For example, it can support fetching costs from different cloud providers based on some annotations (currently I see it is hardcoded to support GCP).

Regarding your questions:

> im assuming the backend isnt setup to just handle query of just a specific gcp cost data and is made available for the api client to use yet. How do you suggest i move forward? Should i be using another existing API instead or just build one specific for gcp?

You are correct. Currently InfraWallet backend `reports` endpoint only handles cloud provider tags (AWS and Azure supported, using the `tags` parameter while `filters` is not used), and then responds with all the cost data (service-level granularity). The aggregation is done by the frontend because for all the aggregations, the frontend does not need to query the data again if the query time range does not change.

I think in your case, you may need to implement this `filters` function for the backend. If so, the current InfraWallet frontend works as usual because it does not use any `filters` and still groups data after fetching all of the cost reports, and your `EntityInfraWalletCard` can send a request with `filters` to the backend to get what you need.

In the current implementation, we pass the `filters` to each provider client and think it may be used when calling cloud provider cost APIs. However, if we would like to filter by `account`, or `provider`, etc., it might be better to handle such `filters` after getting cost reports (after this [line](https://github.com/electrolux-oss/infrawallet/blob/main/plugins/infrawallet-backend/src/service/router.ts#L108))

> also the current flow seems to trigger a new job per each entity that has a gcp project. I think that is uneconomical , can i re-use the first job being ran via the main infrawallet plugin api call, re-use that data for every entity, then only selectively update per entity data only when the main plugin get an update?

InfraWallet backend has used Backstage caching service. I think if backend supports handling the `filters`, it should be good enough for `EntityInfraWalletCard`.

What do you think? Besides, that would be nice if you could share a bit more about your use cases. Do you manage GCP projects using Catalog, or do you list/manage cloud resources in GCP using Catalog?

nia-potato commented:
Thanks for the quick response @gluckzhang 

> I think it can be more generic. For example, it can support fetching costs from different cloud providers based on some annotations (currently I see it is hardcoded to support GCP).

makes sense, however i am not too familiar with AWS or Azure, i've only worked with Alicloud in the past besides GCP, and i heard it is a similar flow with AWS where instead of group projects by project IDs like in GCP, AWS is using RAM users. Azure i remember it has resource manager, but once again i dont know what would be generic field that we can pass independent of the provider.  However we could do it like this: 

```
infrawallet.io/gcp-project-ids: projectIDA, projectIDB
infrawallet.io/aws-$Identifier: $idA $idB 
infrawallet.io/azure-$identifier: $idA $idB. 
```

i can pre-add the logic to read these annotations for the MVP, then however knows what they are doing with what ID it should actually be can change it later. But i can make sure it will be a provider based fetch instead of just fetching from GCP project IDs. 

lmk if that is what you were looking for, or if you were planning something else please lmk 



> I think in your case, you may need to implement this filters function for the backend. If so, the current InfraWallet frontend works as usual because it does not use any filters and still groups data after fetching all of the cost reports, and your EntityInfraWalletCard can send a request with filters to the backend to get what you need.


noted, i will try implementing the backend for this.  


> InfraWallet backend has used Backstage caching service. I think if backend supports handling the filters, it should be good enough for EntityInfraWalletCard.

I am not too familiar with with the Backstage caching service, thus i would like to double check with you on this before i start. 
How do i ensure the new filters backend will be using the caching service? is there existing code that you can link to me as a example? 

Also with the new filters backend that we create soon, how do i prevent the case, where every Entities query to the backend creates a new Bigquery job for the filters api call?  

Although we can assume that the subsequent  filters API call would be reading cached data from backstage, but for companies that have alot of entities that could be alot of jobs initially. 

Is there any way we could do would normal api call that fetches all the data as a single job, the subsequent api call will read on that cached data? instead of triggering everything individually? 

> What do you think? Besides, that would be nice if you could share a bit more about your use cases. Do you manage GCP projects using Catalog, or do you list/manage cloud resources in GCP using Catalog?


Yea of course, so my use case is we have a global backstage google service account in the organization level that has read permission, and then our catalog will read all our gcp projects + gke clusters and auto add these as backstage resources: `gcp-project` and `kubernetes-cluster`, so that users can freely link and adds `dependsOn` resources freely without having to care about auth at all. 

Hence the reason why i suggested the second use case of, if it detects a entity has a `dependsOn` resource and that resource type is `gcp-project` infrawallet frontend will extract that value and resolve that cost of that project. 

This may sound like something pertaining to my environment, but even if people dont use the automatic catalog importer, folks that would like to manage their gcp projects within backstage as resource, will have the `gcp-project` as a resource, whether its being created via Scaffolder flow, or manually created, as long as people want to link it. 

Thus adding the dependsOn logic detection is a more native way for this particular resource, instead of expecting the user to add the annotation. I would hope such logic would be for aws + azure as well but i dont have the knowledge to do so on that front. But of course, this is a dual condition with extracting the data from the annotation as discussed above, for folks that do not have the automated setup for compatibility. 

hope that makes sense.

gluckzhang commented:
Hi @nia-potato, what about having the following (optional) annotations:

<img width="1482" alt="image" src="https://github.com/user-attachments/assets/f6bb64f0-9451-4a2d-bfe4-75b20b5255f7">

Based on the screenshot, in `EntityInfraWalletCard` we can always check the following annotations of a Catalog entity:

```
infrawallet.io/account: "xxx"
infrawallet.io/service: "xxx"
infrawallet.io/category: "xxx"
infrawallet.io/provider: "xxx"
infrawallet.io/project: "xxx"
infrawallet.io/extra-filters: "key-x: value-x, key-y: value-y"
```

The idea is to merge these annotations into one `key-value` string and pass it as a `filters` parameter to the `reports` endpoint.

```
.../api/infrawallet/reports?&filters=(provider:gcp,project:projectIDA)&tags=()&groups=&granularity=monthly&startTime=1711922400000&endTime=1727733599999
```

Then in our backend, after this [line](https://github.com/electrolux-oss/infrawallet/blob/main/plugins/infrawallet-backend/src/service/router.ts?rgh-link-date=2024-09-24T08%3A32%3A30Z#L108), we update the `results` list to filter the reports before the response.

In this way, we reuse the cached cost data for all the `EntityInfraWalletCard` requests because when the first request arrives (`granularity=monthly&startTime=1711922400000&endTime=1727733599999` is needed as well when sending request from `EntityInfraWalletCard`), InfraWallet fetches all the costs data from different cloud providers. Then for all the following requests, the backend reuses the cached data and just filters the reports before a response.

Does it make sense to you?

nia-potato commented:
@gluckzhang Makes sense, thanks for the explanation.  i did a quick mock up (not finalized) 

<img width="734" alt="Screenshot 2024-09-24 at 23 30 54" src="https://github.com/user-attachments/assets/3f5b260f-c29d-4821-9fd8-ba5de7b23bb6">

<img width="732" alt="Screenshot 2024-09-24 at 23 31 20" src="https://github.com/user-attachments/assets/70609f68-3dca-40c4-a2fd-855a08297a66">


but im currently double checking the costs to make sure it is actually correct.

gluckzhang commented:
Thanks @nia-potato for trying out the idea! Yeah, it quite matches what I am thinking :)

Regarding filters, I don't think we need to provide UI in `EntityInfraWalletCard` because for a Catalog entity, it already has the filters defined as annotations and users do not really need to change them.

> there are a few entries like this, 0 cost has a percent increase while the value stays constant.

Thanks for reporting this. I think it is related to the precision. It happens when the number is very small. We will fix it in the next release.

nia-potato commented:
@gluckzhang 
here is the current behavior: 

taking project-ids for example since i have such test data ready, if we set:
```
    infrawallet.io/project: backstage-dev,backstage-prod
```

we get the below chart
<img width="603" alt="Screenshot 2024-09-26 at 00 59 14" src="https://github.com/user-attachments/assets/c2c1f2a5-0af3-4c71-9cf3-a7961e169fc6">
<img width="603" alt="Screenshot 2024-09-26 at 00 59 08" src="https://github.com/user-attachments/assets/2a3abaa1-c3cc-44ba-810b-faf44716484a">


if we set 
```
    infrawallet.io/project: backstage-dev,backstage-prod
    infrawallet.io/category: Compute
```
We get the below: 

<img width="600" alt="Screenshot 2024-09-26 at 01 02 22" src="https://github.com/user-attachments/assets/113fc42b-b5a0-4f6c-94f5-6ad3b0a34ca0">
<img width="599" alt="Screenshot 2024-09-26 at 01 02 34" src="https://github.com/user-attachments/assets/1b3c2977-f733-41a1-aea4-23ef8998f2bc">


same applies for the other fields. If it is all you expect, il will clean up code and push it up.

gluckzhang commented:
Hi @nia-potato, it looks pretty nice, thanks! I have only one concern: currently we are migrating our charts written in [apexcharts](https://apexcharts.com/docs/react-charts/) to [MUI X Charts](https://mui.com/x/react-charts/) because Backstage plugins use MUI for most of the components. Would it be possible for you to use [MUI X Charts](https://mui.com/x/react-charts/) for the charts in `EntityInfraWalletCard`?

It is fine if you don't have time for that. We can merge the PR but later on I might spend some time on rewriting the charts :)

nia-potato commented:
hi @gluckzhang  glad you like it, yea can you first help me merge it and if its not too much to ask, do a quick small release if possible ? i will come back few weeks after to switch the charts to use MUI  

and also add the notification (if cost too high)  part.

gluckzhang commented:
@nia-potato Of course :) Please fix the conflicts and then I will merge the PR. I have added a pipeline that publishes release candidates after merging code into the main branch. So later on you will get a version like https://www.npmjs.com/package/@electrolux-oss/plugin-infrawallet/v/0.1.11-20240926113301-340703c and start using it 🎉

gluckzhang commented:
By the way, version `v0.1.12` will be released maybe next week and `EntityInfraWalletCard` will also be there!

nia-potato commented:
hi @gluckzhang  so sry about this, for some reason after i rebased and fixed the conflicts when i tried to add files for just this repo, i am getting a `failed to execute git` in my IDE since the conflict is just adding `parseFilters` can you help me resolve it? 

i need to go fix my IDE

nia-potato commented:
@gluckzhang  should be all set. Thanks!

gluckzhang commented:
Thanks for the contribution, the PR has been merged! 🎉 
After this job is done, you will be able to find the NPM package that contains this feature here: https://www.npmjs.com/package/@electrolux-oss/plugin-infrawallet?activeTab=versions :)
